File: CSharpCodeBase/components/controlsystem/spells/DeathClickSpell.lua
local Action = require 'components.controlsystem.action'
local Class = require 'utils.hump.class'
local Vector3 = require 'utils.hump.vector3'
local HitTester = require 'utils.hittest'

local STATE_MOVING = 0
local STATE_REQUEST_ATTACK = 1
local STATE_ATTACK = 2

local SpellAction = Class({__includes=Action, init = function (self, name)
      Action.init(self, 3, false, name)
      self.position = nil
end})

function SpellAction:OnStartRunning()    
  -- call spell animation
  if self.position then    
    self.target = Input.RaycastTargetEntityWithRadius(1, self.position)        
  else
    self.target = Input.RaycastTargetEntityWithRadius(1)    
  end
end

function SpellAction:OnStopRunning()
  -- stop spell animation?
end


function SpellAction:Update()
  if not self.target then return true end
  
  if self.target.Death then
    self.target:Death()
  end
  -- cast spell
  return true
end

function SpellAction:OnEvent(data)
  -- Cast spell by animation here
end

function SpellAction:IsContinuous()
  return false
end

function SpellAction:BeginDraw()
end

function SpellAction:StopDraw()
end

function SpellAction.CanApply(position)    
  if position then        
    return Input.RaycastTargetEntityWithRadius(1, position) ~= nil
  else        
    return Input.RaycastTargetEntityWithRadius(1) ~= nil
  end
end

return SpellAction

File: CSharpCodeBase/components/controlsystem/spells/DeathClickSpell.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class DeathClickSpell :action{
 var STATE_MOVING = 0;
 var STATE_REQUEST_ATTACK = 1;
 var STATE_ATTACK = 2;
 var SpellAction = Class({__includes=Action, init = public void (self, name){
       Action.init(self, 3, false, name);
       this.position = null;
 }})
 public void SpellAction:OnStartRunning()    {
   // call spell animation
   if(this.position  ){     
     this.target = Input.RaycastTargetEntityWithRadius(1, this.position)        ;
   }else{
     this.target = Input.RaycastTargetEntityWithRadius(1)    ;
   }
 }
 public void SpellAction:OnStopRunning(){
   // stop spell animation?
 }
 public void SpellAction:Update(){
   if(not this.target  ){  return true }
   
   if(this.target.Death  ){ 
     this.target:Death();
   }
   // cast spell
   return true;
 }
 public void SpellAction:OnEvent(data){
   // Cast spell by animation here
 }
 public void SpellAction:IsContinuous(){
   return false;
 }
 public void SpellAction:BeginDraw(){
 }
 public void SpellAction:StopDraw(){
 }
 public void SpellAction.CanApply(position)    {
   if(position  ){         
     return Input.RaycastTargetEntityWithRadius(1, position) != null;
   }else{        
     return Input.RaycastTargetEntityWithRadius(1) != null;
   }
 }
 return SpellAction}}